Test/Tests/RealtimeClientPresenceTests.swift

```diff
@@ -756,86 +756,106 @@ class RealtimeClientPresenceTests: XCTestCase {
+            mainChannel.presence.enter(nil) { error in
                 XCTAssertNil(error)
                 partialDone()
             }
-            channel.presence.get { members, error in
+            leavesChannel.presence.enter(nil) { error in
+                XCTAssertNil(error)
+                partialDone()
+            }
+            mainChannel.presence.get { members, error in
                 XCTAssertNil(error)
                 if let members {
-                    XCTAssertEqual(members.count, 3)
+                    XCTAssertEqual(members.count, 3) // "user1", "user2", "leaves"
                 } else {
                     XCTFail("Expected members to be non-nil")
                 }
                 partialDone()
             }
-            channel.presence.subscribe(.enter) { message in
-                if message.clientId == "tester" {
-                    channel.presence.unsubscribe()
+            mainChannel.presence.subscribe { message in
+                if message.clientId == "main" {
+                    XCTAssertEqual(message.action, ARTPresenceAction.enter)
+                    mainChannel.presence.unsubscribe()
+                    partialDone()
+                }
+                else if message.clientId == "leaves" {
+                    XCTAssertEqual(message.action, ARTPresenceAction.present) // .enter was replaced by .present (RTP2d)
                     partialDone()
                 }
             }
         }
 
+        var presenceEvents = [ARTPresenceMessage]()
         waitUntil(timeout: testTimeout) { done in
             let partialDone = AblyTests.splitDone(4, done: done)
-            channel.presence.subscribe { presence in
-                XCTAssertEqual(presence.action, ARTPresenceAction.leave)
-                XCTAssertEqual(presence.clientId, "tester")
-                partialDone()
-            }
-            channel.once(.suspended) { _ in
-                channel.internalSync { _internal in
-                    XCTAssertEqual(_internal.presenceMap.members.count, 4)
-                    XCTAssertEqual(_internal.presenceMap.localMembers.count, 1)
+            mainChannel.presence.subscribe { presence in
+                presenceEvents += [presence]
+                delay(1) {
+                    partialDone() // Wait a bit to make sure we don't receive any other presence messages
                 }
-                partialDone()
             }
-            channel.once(.attached) { stateChange in
-                XCTAssertNil(stateChange.reason)
-                channel.presence.leave(nil) { error in
+            mainChannel.once(.suspended) { _ in
+                mainChannel.internalSync { _internal in
+                    XCTAssertEqual(_internal.presenceMap.members.count, 4) // "main", "user1", "user2", "leaves"
+                    XCTAssertEqual(_internal.presenceMap.localMembers.count, 1) // "main"
+                }
+                leavesChannel.presence.leave(nil) { error in
```

Comment:
What makes you sure that this presence leave occurs whilst `mainChannel` is `SUSPENDED`? What if `mainChannel` reattaches before this line’s presence leave completes?

Reply:
https://github.com/ably/ably-cocoa/pull/1793/commits/9a11e4383d16fd72c65554a7a65123283bde0788

Reply:
I don't think that the situation is different in your new version of the code.

You're triggering the `leavesChannel` presence leave when `mainChannel` enters the `SUSPENDED` state. But, upon entering the `SUSPENDED` state, `mainChannel` will trigger a channel attach operation. How do you know that the presence leave will complete before the attach?

I'm not sure of a good way of being sure of this. The ideal would be to make sure that the presence leave operation completes before `mainChannel` tries to reattach. Is there a way, for example, that we could delay the sending of the `ATTACH` `ProtocolMessage` until the presence enter has completed? Perhaps using `TestProxyTransport`?

Reply:
https://github.com/ably/ably-cocoa/pull/1793/commits/e04e5865f0bfc2091aaaa8417e60634baa1aa248